### PR TITLE
Enable rkt driver to use address_mode = 'driver'

### PR DESCRIPTION
**Pull request by dalegaard**
The rkt driver currently does not allow a user to specify address_mode = 'driver' in the service because it doesn't pass a DriverNetwork instance up.

This PR fixes that, by filling out a DriverNetwork instance with the first IP found on the container and generating a matching portmap.

To do this, the UUID of the rkt container must first be tracked correctly. This was done in a pretty racy way before, so here we've also refactored the rkt driver's `Start` method to instead use `prepare` followed by `run-prepared`.

BR.

### CODE REVIEW

File: client/driver/rkt.go [hunk 1]
```diff
@@ -523,8 +588,51 @@ func (d *RktDriver) Start(ctx *ExecContext, task *structs.Task) (*StartResponse,
...
+			}
+
+			driverNetwork = &cstructs.DriverNetwork{
+				PortMap: portmap,
+				IP:      status.Networks[0].IP.String(),
+			}
+			break
+		}
+	}
+	if driverNetwork == nil {
```
Comment (schmichael) [inline client/driver/rkt.go]: In the Docker driver we return an error if `driverConfig.PortMap` is set but no networks could be found. I think the same behavior would be appropriate here.

https://github.com/hashicorp/nomad/blob/master/client/driver/docker.go#L1065-L1070

**Comment by dalegaard** [comment]
@schmichael looks like the tests are failing now because of https://github.com/rkt/rkt/pull/3699 affecting rkt < 1.27.0, and travis is using 1.18.0.

Which fallback would you prefer for this case? I'm guessing it needs to be handled somehow, because else rkt < 1.27.0 will no longer be usable with Nomad, even when the address_mode option is not set to 'driver'.

**Comment by dalegaard** [comment]
@schmichael Sure, no problem - do you want me to add a note in the changelog under a backwards incompatibilities header while I'm at it?

**Comment by schmichael** [comment]
@dalegaard good thinking. You're better at this than me. ;)

**Comment by dalegaard** [comment]
So this is proving a bit more elusive than expected, since the test cases pass fine on my local machine but not in Travis. This prompted me to add some more logging in there, so I guess it's a win overall :-) Sorry for abusing your Travis setup like this ;-)

**Comment by dalegaard** [comment]
@schmichael right, turns out this was failing because the PortsMapping test runs redis with `--version`, making it return before the status has been queried, so a classic race condition.

This does bring up another case however: what to do if the task returns before its status could be queried succesfully? Should we check if the task has already returned, and then ignore the network errors? Is there even a non-racy way to do such a check?

**Comment by schmichael** [comment]
@dalegaard Argh that is unfortunate, and I've definitely run into similar issues before. Perhaps you could move the [post-run status checking code](https://github.com/hashicorp/nomad/pull/3256/files#diff-f5de7f60c289c028cc4fdeadccfe65dfR593) into a method, and if that method returns an error, check if the executor (started in a goroutine right about the linked line) has exited. If it's exited already, there's no need to return an error from `Start()` as the either the task will fail or it will be restarted and `Start()` will get called again.

If the task fails, then the fact that a network wasn't returned won't hurt anything.

If the task is restarted, `Start()` will get another chance to read the network. If the initial fast exit was due to a transient failure, everything may Just Work upon restart and the network will get read. Otherwise the task will eventually fail and the lack of a successful network read won't matter.

Sorry for the verbose explanation, but I'm trying to work through it myself. :smile: The summary is: if the executor/handle exits before you're able to read the network there's no need to worry about returning an error regarding the network. It would only mask the actual failure.

**Comment by dalegaard** [comment]
@schmichael I pushed an updated version where I've refactored out the network information polling loop, per your suggestion. Looks much neater now :-)

The build failed with a "port collision" in an unrelated test case however -.-

**Comment by schmichael** [comment]
@dalegaard Updates look great! I kicked the build because you deserve to see all green checks. Will merge once Travis is happy. Thanks for sticking with this! That was a really tricky problem.

**Comment by schmichael** [comment]
Confirmed with a coworker that failure is unrelated and it's safe to merge as-is. Sorry no greenlight from Travis, but thanks for the hard work @dalegaard!

If you're looking for some simple rkt improvements, #3207 may be a one liner. Testing - as always - will be the tricky part (although moving that validation to a method and unit testing it is totally acceptable instead of a full rkt integration test).

**Comment by dalegaard** [comment]
@schmichael No worries, thanks for reviewing this!

**Pull request merged**